packages/neon_framework/packages/notifications_push_repository/lib/src/notifications_push_repository.dart

```diff
@@ -0,0 +1,243 @@
+        if (subscription == null) {
+          _log.finer('Subscription for $accountID not found'); // coverage:ignore-line
+          return;
+        }
+
+        subscription = subscription.rebuild((b) => b.endpoint = null);
+        subscription = await _unregisterNextcloud(accountID, account, subscription);
+
+        await _saveUpdatedSubscription(subscriptions, accountID, subscription);
+      },
+      onMessage: _onMessage,
+    );
+
+    await _updateSubscriptions();
+
+    _initialized = true;
+  }
+
+  Future<void> _updateSubscriptions() async {
+    if (_selectedDistributor == null) {
+      _log.fine('Push notifications disabled, removing all subscriptions');
+
+      await _unregisterUnifiedPush();
+      return;
+    }
+
+    if (_selectedDistributor != _previousDistributor) {
+      _previousDistributor = _selectedDistributor;
+      _log.finer('UnifiedPush distributor changed to $_selectedDistributor');
+
+      await _unregisterUnifiedPush();
+      await UnifiedPush.saveDistributor(_selectedDistributor!);
+    }
+
+    await _registerUnifiedPush();
+  }
+
+  Future<void> _registerUnifiedPush() async {
+    // Notifications will only work on accounts with app password
+    for (final account in _accounts.where((a) => a.password != null)) {
+      _log.finer('Registering ${account.id} for UnifiedPush');
+
+      await UnifiedPush.registerApp(account.id);
+    }
+  }
+
+  Future<void> _unregisterUnifiedPush() async {
+    final subscriptions = await _storage.readSubscriptions();
+    for (final entry in subscriptions.entries) {
+      final accountID = entry.key;
+      final account = _accounts.firstWhereOrNull((account) => account.id == accountID);
+      var subscription = entry.value;
+
+      subscription = await _unregisterNextcloud(accountID, account, subscription);
+      if (subscription.endpoint != null) {
+        _log.finer('Unregistering $accountID from UnifiedPush');
+
+        subscription = subscription.rebuild((b) => b.endpoint = null);
+        await _saveUpdatedSubscription(subscriptions, accountID, subscription);
+
+        await UnifiedPush.unregister(accountID);
+      } else {
+        await _saveUpdatedSubscription(subscriptions, accountID, subscription);
+      }
+    }
+  }
+
+  Future<PushSubscription> _registerNextcloud(Account account, String endpoint, PushSubscription subscription) async {
+    _log.finer('Registering ${account.id} at Nextcloud');
+
+    try {
+      final response = await account.client.notifications.push.registerDevice(
+        $body: notifications.PushRegisterDeviceRequestApplicationJson(
+          (b) => b
+            ..pushTokenHash = notifications.generatePushTokenHash(endpoint)
+            ..devicePublicKey = _privateKey.publicKey.toFormattedPEM()
+            ..proxyServer = '$endpoint#',
+        ),
+      );
+
+      return subscription.rebuild(
+        (b) => b
+          ..endpoint = endpoint
+          ..pushDevice.replace(response.body.ocs.data),
+      );
```

Comment:
This effectively builds a new subscription (every attribute is set) so I don't see the point of rebuilding an old one over just creating one.

Reply:
The endpoint shouldn't be set here as it is already set before, that's a bug here.

Reply:
The endpoint shouldn't be set again here as it is already set before.

Reply:
Why not return the relevant data (pushDevice) and let the caller build the subscription?
That would also clean up the other method.